### PR TITLE
Build che server Docker image with CentOS base image

### PR DESCRIPTION
**Pull request by dharmit**
### What does this PR do?

Build Che Server Docker image on CentOS using [CentOS Community Container Pipeline](https://github.com/CentOS/container-pipeline-service/).
### New behavior

Che Server image can now be built on top of CentOS base image. 
### PR type
- [x] Minor change = no change to existing features or docs
- [ ] Major change = changes existing features or docs
### Minor change checklist
- [ ] New API required?
- [ ] API updated
- [ ] Tests provided / updated
- [ ] Tests passed
### Major change checklist
- [ ] New API required?
- [ ] API updated
- [ ] Documentation provided (include here or link to docs)
- [ ] Tests provided / updated
- [ ] Tests passed

Please review [Che's Contributing Guide](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md) for best practices.

### CODE REVIEW

**Comment by dharmit** [comment]
cc @l0rd

File: Dockerfile.centos
Comment (benoitf) [inline Dockerfile.centos]: as this Dockerfile is not using current assembly binary but downloading it, should this PR be moved to https://github.com/eclipse/che-dockerfiles ?

File: Dockerfile.centos
Comment (benoitf) [inline Dockerfile.centos]: to avoid several layers, RUN instructions should be merged into one (and tar command should be for example piped to curl (no need to remove at the end, etc)

**Review by l0rd — COMMENTED** [review]
@benoitf @TylerJewell CentOS container pipeline is already setup to build this image. As a result we can now:

``` sh
docker pull http://registry.centos.org/eclipse/che-server
docker pull http://registry.centos.org/eclipse/che-launcher
```

Even if the Dockerfile is not 100% completed (the assembly binary is downloaded and not copied form local folder) we thought it would be better to move it from @dharmit fork to here. What do you think?

**Comment by TylerJewell** [comment]
Can you explain the connection to openjdk and updating this dockerfile to have a build context similar to the default one we currently have?  I don't see the dependency.

The build context just calls for an assembly directory to be present and added into the built image.  The dockerfile shouldn't have an opinion about how that assembly wells previously built.

**Comment by l0rd** [comment]
@TylerJewell we are splitting the CI for CentOS container pipeline in 3 steps:

1) Build of Che from source code (with OpenJDK and nodejs 4 or 6)
2) Build the che-server Docker image and publish it 
3) Deploy Che on an OpenShift cluster for testing 

We decided to start with steps  2) and 3) and leave 1) as last step. That's the reason why we download the binary: we don't have an automated build that generates it.

By the way the more I think about that the more I think we should provide the right Dockerfile from the beginning. It does not make sense to commit a Dockerfile that we know we are going to change in a couple of weeks. We should modify Dockerfile.centos to copy the assembly binary from the local folder. And implement a workaround in our CI until step 1) is completed. A couple of possible workarounds:
- build Che with current default tools/version (i.e. Oracle JDK and nodejs012) or
- download the binary in folder /assembly/assembly-main/ before running `docker-build` 
  @dharmit what do you think? Would it be possible for you do these changes?

**Comment by TylerJewell** [comment]
Yes, that was my request as well.  Let's leave the Dockerfile with the syntax of using local assembly.  So I think both of your options are suitable as a way to solve your goals.  And then this way the Dockerfile for different OS can remain consistent.  I support that as part of this, we should rename the base Dockerfile to also have a .ubuntu extension?

File: cccp.yml
Comment (dharmit) [reply cccp.yml]: @l0rd Yes, this is required to build the image in CentOS pipeline. It's referenced [here](https://github.com/CentOS/container-index/blob/master/index.d/eclipse.yml#L70).

**Comment by benoitf** [comment]
@TylerJewell base Dockerfile is using alpine so .ubuntu extension is not a candidate. so you mean Dockerfile.alpine ? BTW I think it's easier to have minimal image (based on alpine) using name Dockerfile (this name is also handled in editors, etc for syntax highlighting)

File: Dockerfile.centos
Comment (benoitf) [inline Dockerfile.centos:12]: do we need to have registry.centos.org/ when trying to build the image ?

Comment (dharmit) [reply Dockerfile.centos:12]: @benoitf You don't need to have it. It's already up and running on the CentOS infra.

File: Dockerfile.centos
Comment (dharmit) [reply Dockerfile.centos:23]: They're the same.

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:12]: I mean, in the comment, why do we need to specify the registry prefix, as user wanted to try doesn't need it

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:23]: why we don't specify FROM centos:latest then ?

File: Dockerfile.centos
Comment (dharmit) [reply Dockerfile.centos:12]: Right, if it's being built locally, a user probably doesn't need it. But for che-launcher image to be able to find it properly on local setup, it would be a good idea to name/tag it that way. Let me know if that understanding is incorrect. 

Also, I've taken a cue from the build statement [here](https://github.com/eclipse/che/blob/master/Dockerfile#L12).

File: Dockerfile.centos
Comment (dharmit) [reply Dockerfile.centos:23]: Because the registry mentioned here is local to the pipeline. That helps reduce the build time and saves network resources as well.

**Comment by TylerJewell** [comment]
Can you explain what it means to be local to the pipeline, please?  I am not understanding that syntax and the differences between having a fully explicit registry vs the default centos in the FROM.

Florent - you are right about alpine.  So dockerfiles with alpine get no extension and dockerfiles from other base images do get an extension.

I noticed that haproxy and nginx images once tagged append -alpine to the version tag for alpine images and nothing for other images types.

**Comment by dharmit** [comment]
@TylerJewell 

> Can you explain what it means to be local to the pipeline, please? I am not understanding that syntax and the differences between having a fully explicit registry vs the default centos in the FROM.

It's in the local network. Registry and the systems on which images get built are in the same network. 

As far as the differences are concerned, CentOS registry is working on providing features of scanning and testing the images before they're pushed to registry. Also, registry.centos.org is canonical upstream for container workloads for CentOS project.

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:23]: @dharmit but for users wanted to use this Dockerfile, they are not local to the pipeline.

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:12]: on a local setup, there is a way to specify the name of the local image for example

Comment (benoitf) [inline Dockerfile.centos]: Else it would act for me, like storing in eclipse/che repository a file that is internal to registry.centos.org and use privately

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:23]: Else it would act for me, like storing in eclipse/che repository a file that is internal to registry.centos.org and used privately

Comment (dharmit) [reply Dockerfile.centos:23]: @benoitf 

> but for users wanted to use this Dockerfile, they are not local to the pipeline

Agreed. Instead of using the Docker hub, it's pulling from registry.centos.org.

> Else it would act for me, like storing in eclipse/che repository a file that is internal to registry.centos.org and used privately

Nothing's getting used privately. Anyone can pull from registry.centos.org. Also, anyone can build the image using the exact same Dockerfile on their local system.

If that doesn't answer your question, I'm not sure I understand your concern here.

Comment (l0rd) [reply Dockerfile.centos:23]: @benoitf @TylerJewell  registry.centos.org is a public registry of Docker images (as the Docker hub). Since it's maintained from CentOS team it's a better source for CentOS base images than docker hub is (e.g. images there are available earlier). Plus we are going to publish che-server:xx-centos on registry.centos.org so it makes more sense to use it as source in this case.

Comment (TylerJewell) [reply Dockerfile.centos:23]: My question about what you wanted to do with the pipeline makes sense.  I think the only fixes remaining are:
1. Adjust the Dockerfile so that it pulls from a local assembly that is built instead of downloading from nightly the assembly.
2. Comine the RUN commands into a single RUN command to remove layers.

Then we should be ok to merge.

Comment (dharmit) [reply Dockerfile.centos:23]: @l0rd @TylerJewell does that look OK now?

File: Dockerfile.centos
Comment (benoitf) [inline Dockerfile.centos]: as you're installing "java" with yum
how can you be sure that JAVA_HOME will be installed in this folder ?
if there are updates on JAVA in centos latest image, it will break, no ?

File: Dockerfile.centos
Comment (benoitf) [inline Dockerfile.centos]: if you install ca-certificates and curl and openssl to only download docker, it should be removed after docker has been downloaded to reduce image size.

File: Dockerfile.centos
Comment (benoitf) [inline Dockerfile.centos]: this ENV instruction can be added to the existing ENV command at the top of the file

File: cccp.yml
Comment (benoitf) [reply cccp.yml]: I don't see why a file needs to be added in a repository for building it on a CI server.
If each CI needs to add config files in the repository, it will be a mess.

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:12]: I still don't get why default image would be registry.centos.org/eclipse/che-server in this instruction as we can build on our own and then it will be local and we could use it on any registry. So registry part is not required

Comment (l0rd) [reply Dockerfile.centos:12]: @benoitf I think that's to be consistent with the comments on other Dockerfiles https://github.com/eclipse/che/blob/master/Dockerfile#L21. The registry and repository correspond to the ones used by the CI. And that's still a command that a user can execute safely on its machine. If a user want to push the resulting image (to the docker hub or to another registry) he will need to re-tag it anyway with his <registry>/<repo>/che-server.

Comment (benoitf) [reply Dockerfile.centos:12]: to be more explicit. Dockerfile.centos is producing a drop-in replacement for Dockerfile. so the image name produced should be the same.
It's the CI that may deploy it on its own registry by adding the registry prefix.
But all Dockerfiles should produce the same image name. As they're alternatives.

File: Dockerfile.centos
Comment (l0rd) [reply Dockerfile.centos]: @benoitf I agree with you on that but I'm not sure there is a better option. @dharmit can we specify the `JAVA_HOME` path in centos without being so explicit about the JDK version? Like [it's done for the alpine image](https://github.com/eclipse/che/blob/master/Dockerfile#L26).

File: Dockerfile.centos
Comment (kbsingh) [inline Dockerfile.centos]: is there a reason why we are not using the distro provided docker here ? that would better line up with the entire stack, both for buildtime and runtime.

**Review by kbsingh — CHANGES_REQUESTED** [review]

File: Dockerfile.centos
Comment (TylerJewell) [reply Dockerfile.centos]: I think it has to do with limitations in how they distribute Docker 1.6, which uses a different location and distro than others.

Comment (kbsingh) [reply Dockerfile.centos]: could you point me at something that quantifies the need to remain with docker 1.6 ?

Comment (TylerJewell) [reply Dockerfile.centos]: We use docker 1.6 in our images which are responsible for creating other images to avoid having incompatibilities with different daemons.  The docker daemon is installed by users outside of our control and can vary in version and configuration.  We do the same in the Che launcher image too.

Comment (kbsingh) [reply Dockerfile.centos]: ack, thanks. I'll find a more appropriate venue for the 1.6 convesation; but to summarise present state : docker1.6 is used, in order to enable the widest possible user deployment stack selection.

Comment (TylerJewell) [reply Dockerfile.centos]: Correct, yes.  If you are interested I can tomorrow dig up the old issues where some of the inconsistencies were uncovered that led us to this approach.

**Review by l0rd — CHANGES_REQUESTED** [review]
@dharmit I've built and tested your Dockerfile and it works like a charm :-) However I think we still need to solve the following points to be able to merge:
- Move the declaration of environment variable `CHE_HOME` with the previous `ENV` declarations
- Remove some package (like curl) as soon as they are not needed anymore
- Look if it's possible to specify `JAVA_HOME` without hardcoding the java version
- Look at a solution for file `cccp.yml`. That's a file that is needed for CentOS pipeline. From the point of view of a developer this file is probably useless and confusing. And it looks to me that it provides the `job id` only and that this is something the CentOS pipeline already knows. Are we sure we can't avoid it? Would it be possible to put it in a distinct repository?

File: Dockerfile.centos
Comment (benoitf) [reply Dockerfile.centos:12]: Maybe I wasn't clear enough but I didn't talk about pushing it to a repository. To use it locally, you don't need to push on a registry, just build it. And then you can still have a name without registry part.

**Comment by dharmit** [comment]
@l0rd 

> Move the declaration of environment variable CHE_HOME with the previous ENV declarations

done.

> Remove some package (like curl) as soon as they are not needed anymore

can't remove `curl` and `ca-certificates` as they're dependencies for packages like `yum`, etc.

> Look if it's possible to specify JAVA_HOME without hardcoding the java version

done.

> Look at a solution for file cccp.yml. That's a file that is needed for CentOS pipeline. From the point of view of a developer this file is probably useless and confusing. And it looks to me that it provides the job id only and that this is something the CentOS pipeline already knows. Are we sure we can't avoid it? Would it be possible to put it in a distinct repository?

I've added more info to the file (as comments) to better explain its purpose.

**Comment by benoitf** [comment]
@dharmit 

> can't remove curl and ca-certificates as they're dependencies for packages like yum, etc.

I don't see why you couldn't remove them as you install curl and ca-certificates few lines before with yum so yum can work without these packages and they're optional

**Comment by dharmit** [comment]
@benoitf removed them from the `yum install` command since those packages are pre-installed.

**Comment by dharmit** [comment]
@l0rd I've added additional info to `cccp.yml` explaining its purpose. Can you please review?

**Comment by TylerJewell** [comment]
@dharmit @l0rd - what are modern day best practices for the files that should go into the root of a repository?  I'm asking because I'm looking at the root of Che repository.  I see the following:
1. Hidden . files, which are used by command-line systems.
2. Markdown files for root-level documentation
3. License, which is mandated by Eclipse Foundation
4. Maven pom for building
5. Che CLI for start / stop / update
6. And then we are starting to have some other items:

`Chefile &`Vagrantfile`, which is really about simplifying getting started with the product, but ultimately not essential for the core product.  And if we introduce the`cccp.yml file` then it's another flavor of this.  I'm wondering - is there a way where Chefile, Vagrantfile, and cccp.yml could be saved as dotfiles for cleanliness purposes?

**Review by l0rd — COMMENTED** [review]
LGTM

**Review by l0rd — APPROVED** [review]

**Comment by l0rd** [comment]
@TylerJewell I think we can rename `cccp.yml` to `.cccp.yml` but I wouldn't dot-prefix Vagrantfile.

**Comment by TylerJewell** [comment]
+1 to the renaming idea.

**Comment by dharmit** [comment]
@l0rd renamed, and did `s/build/include`.

**Comment by l0rd** [comment]
@dharmit we are ready to merge here but you need to sign [the CLA](https://www.eclipse.org/legal/CLA.php) before. And once #2786 get merged we will need to adapt CentOS Dockerfile too.

**Comment by maxandersen** [comment]
> @dharmit we are ready to merge here but you need to sign the CLA before. And once #2786 get merged we will need to adapt CentOS Dockerfile too.

The ip validation (https://dev.eclipse.org/eclipse-webhook/services/status_details.php?id=580461f6bb47e) complains about emails not listed in the commit. Are you sure the ip validation have been rerun against the latest push ?

**Comment by dharmit** [comment]
@maxandersen 

> Are you sure the ip validation have been rerun against the latest push ?

Nope, I'm not sure if they've rerun against the latest push. I'm seeing that since first time I pushed.

**Comment by TylerJewell** [comment]
It has been buggy for awhile.  As long as a pr submitter tells us in the thread that they have executed the cla, we have permission to merge.

**Comment by maxandersen** [comment]
You can use https://projects.eclipse.org/user/cla/validate to check the email too and it does validate dharmit@redhat.com as having signed CLA.

**Comment by TylerJewell** [comment]
@l0rd has commit rights - so will let him as maintainer drive merge, branch deletion, and selection of milestone.

**Comment by benoitf** [comment]
@maxandersen @TylerJewell the CLA check is not buggy.
The commit provided is having "Author: Dharmit Shah shahdharmit@gmail.com"
while CLA is being for dharmit@redhat.com 
so this PR shouldn't have been merged until commit author have been set to dharmit@redhat.com

**Comment by maxandersen** [comment]
@benoitf okey so I now learned to not trust github UI since it does not show the actual author. Note taken! I 

@dharmit you should remember to set the right author not just the signed-off part.

**Pull request merged**
